Security Analysis (#55)
dashpay/dips: dip-0008.md

be able to disrupt the network. Specifically, we will calculate the probability
that an attacker will be able to prevent all ChainLocks for a twelve hour period, and
we will calculate the probability that an attacker can successfully create a malicious
or erroneous ChainLock.

Quorum selection is designed to provide for an effectively simple random
choice of masternodes in each quorum. Thus we may assume that each quorum is
a simple random selection. If there are a total of 'N' masternodes the
number of possible quorums is the number of ways that 400 objects can be
chosen out of `N`. In some literature this number is
written <sub>`N`</sub>`C`<sub>`400`</sub> and is enunciated as
'N choose 400'. Numbers that arise in this fashion are all
binomial coefficients and have a delightful relation to Pascal's Triangle.

For any process with an observable outcome the set of all outcomes is
called a sample space.  If all outcomes are equally likely, as in the
selection of a quorum, the probability of any event is the number of
outcomes that satisfy that event over the total number outcomes.

An attacker can temporarily prevent ChainLocks if they control

> 160 out of 400 masternodes in the quorum.  If the attacker controls

Reviewer: Wouldn't they actually need to control 161? 240/400 could still construct a signature.

Author: Well ... yes.  Good catch.

Author: commit 1f622c903aefff6cfc8df6dcab202e74bf59f05a should address this.  

In a previous commit I restructured the table and the calculations were changed there.

`m` out of the `N` total masternodes we can compute the
number of outcomes that satisfy the event "exactly 160
nodes of the quorum are controlled by the masternode."
Specifically there are

> <sub>`160`</sub>`C`<sub>`m`</sub> * <sub>`240`</sub>`C`<sub>N - m</sub>

Reviewer 2: This looks inconsistent with `<sub>`N`</sub>`C`<sub>`400`</sub>` notion described above. Shouldn't this be 
```suggestion
<sub>`m`</sub>`C`<sub>`160`</sub> * <sub>`N - m`</sub>`C`<sub>240</sub>
```
instead?

(same for the gif below)

Author: Sometimes I get these things confused.  I put 6 choose 3 into google and got 20 so it's always big choose small.

Author: I'll commit this suggestion and then do a full review of every C in the document, including the gif. 

Thanks.

outcomes where the attacker controls exactly `160` masternodes.
Thus, the probability that the attacker controls exactly `160` nodes is

<p align="center"><img src="dip-0008/ThreshCase.gif" alt="Threshold Probability"></p>

> The attacker could also be successful if they control <b>more</b> that `160`

Reviewer 2: ```suggestion
The attacker could also be successful if they control <b>more</b> than `160`
```

of the nodes in the quorum selection. Thus the exact probability that an
attacker can successfully prevent ChainLocks for that quorum round is
computed by counting the number of outcomes where the attacker controls
160 or more nodes in the quorum divided by the total number of outcomes.
This is computed by the sum

<p align="center"><img src="dip-0008/fin_sum.gif" alt="Final sum" ></p> 
  
The numbers involved in this sum can be quite large. We may
use a <A HREF="dip-0008/quorum_attack.py">Python Script</A> to effectively
evaluate this sum. Note
that we made the necessary modifications to this formula
to calculate the probability that an attacker controls 60% or more
of the quorum.

> | N | m |Type of Attack|success probability in one trial |

Reviewer 2: It would probably be better to "squash" the data and restructure this table in another way:
```suggestion
| N | m | Success probability of withholding a ChainLock | Success probability of creating a malicious ChainLock |
```
i.e. the table itself would look like this:
```
| 5000 | 500 | 2.79e-64 | 7.11e-157 |
| 5000 | 1000 | 5.09e-22 | 2.86e-76 |
| 5000 | 1500 | 5.59e-6 | 1.89e-38 |
| 2000 | 200 | 0 | 0 |
| 2000 | 400 | 6.88e26 | 9.49e-94 |
| 2000 | 600 | 1.12e-6 | 3.93e-45 |
```
which IMO is easier to grasp.

Author: I like this and I consider this clearer.

Author: I need to check the math.  Now that I'm looking at it the change of withholding a chainlock with 200 out of 2000 nodes is non zero, but less that 10e-25

Author: *chance of withholding

|---|---|--------------|--------------------------------|
|5000|500| withhold |  2.79e-64|
|5000|1000| withhold | 5.09e-22 |
|5000|1500| withhold | 5.59e-6 |
|5000|500| malicious | 7.11e-157 |
|5000|1000| malicious | 2.86e-76 |
|5000|1500| malicious | 1.89e-38 |
|2000|200 | withhold | 0 |
|2000|400 | withhold | 6.88e26 |
|2000|600 | withhold | 1.12e-6 |
|2000|200 | malicious | 0 |
|2000|400 | malicious | 9.49e-94 |
|2000|600 | malicious | 3.93e-45 |

In all scenarios above, an attacker should not expect to withhold a
ChainLock in the next century. The attacker would have a less than
one in 100 trillion chance of producing at least

> one malicious ChainLock in the next sextillion (10^21) years.

Reviewer 3: assume when we're saying malicious chainlock we're referencing double spend?  Satoshi paper you reference was pretty clear on that maybe you want to spell that out here

Author: I'll look in to making this more clear.  

I'll add a definition of a malicious ChainLock as a chain lock that conflicts with an already produced ChainLock or a ChainLock that introduce a unintended bias.

Author: d64e028  adds a paragraph to explain what could go wrong with a malicious chainlock.  Without mitigation, it could cause a fork, or could introduce undo advantage as with selfish mining.

From these calculations we conclude that an attacker with control of fewer
than 30% of all masternodes will not have any influence on ChainLocks. For
completeness we note that
if an attacker controlled 50% of masternodes an attacker would
generally be able to prevent ChainLocks, but there is less than 30 out of
one billion chance that the attacker will be able to produce at least one

> malicious ChainLock in the next million years.

Reviewer: These numbers assume the most optimistic scenario that the quorum is constructed with a full 400 participating nodes (which is actually pretty close to how things are working). The other end of the spectrum is limited by the [`quorumMinSize` parameter](https://github.com/thephez/dips/blob/master/dip-0006.md#current-llmq-types). For the large LLMQ used by ChainLocks, the allowable number of participating members is 300-400. If the quorum only had 300 members, the threshold for withholding a ChainLock is substantially lower.

Author: I appreciate these comments and will make edits so that this is clear to the reader.

Author: I alerted the reader that this is the case with a new paragraph in 1f622c903aefff6cfc8df6dcab202e74bf59f05a

## Security Considerations

It is of note that it is not required for an attacker to actually own the
collateral of masternodes that the attacker controls. A company offering
the service of hosting masternodes for clients could be persuaded to run

> malicious software. Also if node opperators do not check the digital signature

Reviewer 3: typo on operators

Author: Addressed in 
d64e028

If the need arises to include block signatures in initial block download, we
will update this DIP and implementations accordingly.

## Calculations

> We consider the scenario where an attacker has assumed control of a number of masternodes. We will then calculate two probabilities that the attacker will

Reviewer 2: nit: break in 2 lines
```suggestion
We consider the scenario where an attacker has assumed control of a number of
masternodes. We will then calculate two probabilities that the attacker will
```

selection of a quorum, the probability of any event is the number of
outcomes that satisfy that event over the total number outcomes.

An attacker can temporarily prevent ChainLocks if they control
`161` out of 400 masternodes in the quorum.  If the attacker controls
`m` out of the `N` total masternodes we can compute the
number of outcomes that satisfy the event "exactly `161`
nodes of the quorum are controlled by the masternode."
Specifically there are
<sub>`m`</sub>`C`<sub>`161`</sub> * <sub>`N - m`</sub>`C`<sub>`239`</sub>
 outcomes where the attacker controls exactly `161` masternodes.
Thus, the probability that the attacker controls exactly `161` nodes is

<p align="center"><img src="dip-0008/ThreshCase.gif" alt="Threshold Probability"></p>

The attacker could also be successful if they control <b>more</b> than `160`
of the nodes in the quorum selection. Thus the exact probability that an
attacker can successfully prevent ChainLocks for that quorum round is
computed by counting the number of outcomes where the attacker controls
160 or more nodes in the quorum divided by the total number of outcomes.
This is computed by the sum

> <p align="center"><img src="dip-0008/fin_sum.gif" alt="Final sum" ></p>

Reviewer 2: Can't comment on the gif itself but it looks like the right top part of the equation is "sub" when it shouldn't be:
in the gif: <sub>`m`</sub>`C`<sub>`160 + j` * <sub>`N - m`</sub>`C`<sub>`240 - j`</sub></sub> 
should be: <sub>`m`</sub>`C`<sub>`160 + j`</sub> * <sub>`N - m`</sub>`C`<sub>`240 - j`</sub>

Author: The comment is correct.  The C should not be a subscript.

Author: Corrected in cf01b58

outcomes that satisfy that event over the total number outcomes.

An attacker can temporarily prevent ChainLocks if they control
`161` out of 400 masternodes in the quorum.  If the attacker controls
`m` out of the `N` total masternodes we can compute the
number of outcomes that satisfy the event "exactly `161`
nodes of the quorum are controlled by the masternode."
Specifically there are
<sub>`m`</sub>`C`<sub>`161`</sub> * <sub>`N - m`</sub>`C`<sub>`239`</sub>

> outcomes where the attacker controls exactly `160` masternodes.

Reviewer: One more 160
```suggestion
 outcomes where the attacker controls exactly `161` masternodes.
```

Author: Thanks!

behaving as expected.

It is always recommended that users check the signature on any software they
download. As long as the correct public key is acquired then it is

> cryptographicly not possible for the official download to be spoofed.

Reviewer: ```suggestion
cryptographically not possible for the official download to be spoofed.
```
